File: game-ai/src/main/kotlin/com/thevaguebox/picpac/ai/StochasticMctsAgent.kt
package com.thevaguebox.picpac.ai

import com.thevaguebox.picpac.core.Cell
import com.thevaguebox.picpac.core.Symbol
import com.thevaguebox.picpac.core.ai.AiAgent
import com.thevaguebox.picpac.core.ai.AiDecision
import com.thevaguebox.picpac.core.ai.AiDiagnostics
import com.thevaguebox.picpac.core.ai.AiObservation
import com.thevaguebox.picpac.core.ai.PicPacChanceState
import com.thevaguebox.picpac.core.ai.PicPacDecisionState
import com.thevaguebox.picpac.core.ai.PublicPicPacSearchModel
import com.thevaguebox.picpac.core.ai.SearchLimits
import com.thevaguebox.picpac.core.ai.SearchTransition
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.ensureActive
import kotlin.math.ln
import kotlin.math.sqrt
import kotlin.random.Random
import kotlin.system.measureNanoTime

class StochasticMctsAgent(
    private val defaultSimulations: Int = 1_000,
    private val exploration: Double = sqrt(2.0),
    private val random: Random = Random.Default,
) : AiAgent {
    init { require(defaultSimulations > 0) }

    override suspend fun chooseMove(observation: AiObservation, limits: SearchLimits): AiDecision {
        val rootState = PublicPicPacSearchModel.from(observation)
        val root = DecisionNode(rootState, rootSign = 1)
        val simulations = limits.nodeBudget ?: defaultSimulations
        var completed = 0
        var createdNodes = 1L
        var maxDepth = 0
        val elapsed = measureNanoTime {
            repeat(simulations) {
                currentCoroutineContext().ensureActive()
                if (limits.deadlineNanos?.let { deadline -> System.nanoTime() >= deadline } == true) return@measureNanoTime
                val result = simulate(root)
                createdNodes += result.created
                if (result.depth > maxDepth) maxDepth = result.depth
                completed++
            }
        }
        val best = root.edges.values.sortedWith(
            compareByDescending<ActionEdge> { it.visits }
                .thenByDescending { it.mean }
                .thenBy { moveRank(it.cell) },
        ).firstOrNull() ?: return AiDecision(rootState.legalCells.first())
        return AiDecision(
            best.cell,
            AiDiagnostics.Search(
                nodes = createdNodes,
                cacheHits = 0,
                maxDepth = maxDepth,
                elapsedNanos = elapsed,
                value = best.mean,
                simulations = completed,
            ),
        )
    }

    private fun simulate(root: DecisionNode): SimulationResult {
        val path = arrayListOf<Stats>(root)
        var node = root
        var depth = 0
        var created = 0L
        while (true) {
            depth++
            val edge = if (node.untried.isNotEmpty()) {
                val cell = node.untried.removeAt(random.nextInt(node.untried.size))
                ActionEdge(cell).also {
                    node.edges[cell] = it
                    created++
                }
            } else {
                select(node)
            }
            path += edge

            when (val transition = PublicPicPacSearchModel.place(node.state, edge.cell)) {
                SearchTransition.Win -> return finish(path, node.rootSign.toDouble(), depth, created)
                SearchTransition.Draw -> return finish(path, 0.0, depth, created)
                is SearchTransition.Chance -> {
                    val symbol = sample(transition.state)
                    val child = edge.children[symbol]
                    if (child == null) {
                        val next = DecisionNode(
                            PublicPicPacSearchModel.draw(transition.state, symbol),
                            -node.rootSign,
                        )
                        edge.children[symbol] = next
                        path += next
                        created++
                        val rollout = rollout(next.state, next.rootSign, depth + 1)
                        return finish(path, rollout.reward, rollout.depth, created)
                    }
                    node = child
                    path += node
                }
            }
        }
    }

    private fun select(node: DecisionNode): ActionEdge {
        val logParent = ln(node.visits.coerceAtLeast(1).toDouble())
        return node.edges.values.maxBy { edge ->
            if (edge.visits == 0) Double.POSITIVE_INFINITY
            else node.rootSign * edge.mean + exploration * sqrt(logParent / edge.visits)
        }
    }

    private fun rollout(initial: PicPacDecisionState, initialSign: Int, startingDepth: Int): Rollout {
        var state = initial
        var sign = initialSign
        var depth = startingDepth
        while (true) {
            val winning = state.legalCells.filter {
                PublicPicPacSearchModel.place(state, it) is SearchTransition.Win
            }
            val cell = if (winning.isNotEmpty()) winning[random.nextInt(winning.size)]
            else state.legalCells[random.nextInt(state.legalCells.size)]
            when (val transition = PublicPicPacSearchModel.place(state, cell)) {
                SearchTransition.Win -> return Rollout(sign.toDouble(), depth)
                SearchTransition.Draw -> return Rollout(0.0, depth)
                is SearchTransition.Chance -> {
                    state = PublicPicPacSearchModel.draw(transition.state, sample(transition.state))
                    sign = -sign
                    depth++
                }
            }
        }
    }

    private fun sample(chance: PicPacChanceState): Symbol =
        if (random.nextInt(chance.total) < chance.remainingX) Symbol.X else Symbol.O

    private fun finish(path: List<Stats>, reward: Double, depth: Int, created: Long): SimulationResult {
        path.forEach {
            it.visits++
            it.valueSum += reward
        }
        return SimulationResult(depth, created)
    }

    private fun moveRank(cell: Cell): Int = when (cell.index) {
        4 -> 0
        0 -> 1
        2 -> 2
        6 -> 3
        8 -> 4
        else -> 5 + cell.index
    }

    private interface Stats {
        var visits: Int
        var valueSum: Double
    }

    private class DecisionNode(val state: PicPacDecisionState, val rootSign: Int) : Stats {
        override var visits = 0
        override var valueSum = 0.0
        val untried = state.legalCells.toMutableList()
        val edges = linkedMapOf<Cell, ActionEdge>()
    }

    private class ActionEdge(val cell: Cell) : Stats {
        override var visits = 0
        override var valueSum = 0.0
        val children = hashMapOf<Symbol, DecisionNode>()
        val mean: Double get() = if (visits == 0) 0.0 else valueSum / visits
    }

    private data class Rollout(val reward: Double, val depth: Int)
    private data class SimulationResult(val depth: Int, val created: Long)
}

File: ios/Packages/PicPacKit/Sources/PicPacAI/StochasticMCTSAgent.swift
import Darwin
import PicPacCore

public actor StochasticMCTSAgent: AiAgent {
    public static let defaultSimulations = 1_000
    public static let productionSimulations = 2_000

    private let simulations: Int
    private let exploration: Double
    private var random: any BoundedRandomSource

    public init(
        defaultSimulations: Int = StochasticMCTSAgent.defaultSimulations,
        exploration: Double = sqrt(2)
    ) {
        precondition(defaultSimulations > 0)
        precondition(exploration >= 0 && exploration.isFinite)
        simulations = defaultSimulations
        self.exploration = exploration
        random = SystemAIRandomSource()
    }

    public init<Random: BoundedRandomSource>(
        defaultSimulations: Int = StochasticMCTSAgent.defaultSimulations,
        exploration: Double = sqrt(2),
        random: Random
    ) {
        precondition(defaultSimulations > 0)
        precondition(exploration >= 0 && exploration.isFinite)
        simulations = defaultSimulations
        self.exploration = exploration
        self.random = random
    }

    public func chooseMove(
        for observation: AiObservation,
        limits: SearchLimits = SearchLimits()
    ) async throws -> AiDecision {
        let rootState = PublicPicPacSearchModel.from(observation)
        let root = DecisionNode(state: rootState, rootSign: 1)
        let simulationBudget = limits.nodeBudget ?? simulations
        var completed = 0
        var createdNodes = 1
        var maxDepth = 0
        let clock = ContinuousClock()
        let start = clock.now

        for _ in 0..<simulationBudget {
            try Task.checkCancellation()
            if limits.isPastDeadline { break }
            let result = try simulate(root)
            createdNodes += result.created
            maxDepth = max(maxDepth, result.depth)
            completed += 1
        }

        guard let best = bestFinalEdge(root) else {
            return AiDecision(cell: rootState.legalCells[0])
        }
        return AiDecision(
            cell: best.cell,
            diagnostics: .search(
                SearchDiagnostics(
                    nodes: createdNodes,
                    cacheHits: 0,
                    maxDepth: maxDepth,
                    elapsedNanoseconds: elapsedNanoseconds(start.duration(to: clock.now)),
                    value: best.mean,
                    simulations: completed
                )
            )
        )
    }

    private func simulate(_ root: DecisionNode) throws -> SimulationResult {
        var path: [Stats] = [root]
        var node = root
        var depth = 0
        var created = 0

        while true {
            depth += 1
            let edge: ActionEdge
            if !node.untried.isEmpty {
                let index = try random.nextInt(upperBound: node.untried.count)
                let cell = node.untried.remove(at: index)
                edge = ActionEdge(cell: cell)
                node.edges[cell] = edge
                node.edgeOrder.append(cell)
                created += 1
            } else {
                edge = select(node)
            }
            path.append(edge)

            switch try PublicPicPacSearchModel.place(node.state, at: edge.cell) {
            case .win:
                return finish(
                    path: path,
                    reward: Double(node.rootSign),
                    depth: depth,
                    created: created
                )
            case .draw:
                return finish(path: path, reward: 0, depth: depth, created: created)
            case let .chance(chance):
                let symbol = try sample(chance)
                if let child = edge.children[symbol] {
                    node = child
                    path.append(child)
                } else {
                    let next = DecisionNode(
                        state: try PublicPicPacSearchModel.draw(chance, symbol: symbol),
                        rootSign: -node.rootSign
                    )
                    edge.children[symbol] = next
                    path.append(next)
                    created += 1
                    let rollout = try rollout(
                        initial: next.state,
                        initialSign: next.rootSign,
                        startingDepth: depth + 1
                    )
                    return finish(
                        path: path,
                        reward: rollout.reward,
                        depth: rollout.depth,
                        created: created
                    )
                }
            }
        }
    }

    private func select(_ node: DecisionNode) -> ActionEdge {
        let logParent = log(Double(max(node.visits, 1)))
        var selected: ActionEdge?
        var selectedValue = -Double.infinity
        for cell in node.edgeOrder {
            guard let edge = node.edges[cell] else { continue }
            let value = edge.visits == 0
                ? Double.infinity
                : Double(node.rootSign) * edge.mean
                    + exploration * sqrt(logParent / Double(edge.visits))
            if value > selectedValue {
                selected = edge
                selectedValue = value
            }
        }
        return selected!
    }

    private func rollout(
        initial: PicPacDecisionState,
        initialSign: Int,
        startingDepth: Int
    ) throws -> Rollout {
        var state = initial
        var sign = initialSign
        var depth = startingDepth

        while true {
            let winning = try state.legalCells.filter {
                try PublicPicPacSearchModel.place(state, at: $0) == .win
            }
            let candidates = winning.isEmpty ? state.legalCells : winning
            let cell = candidates[try random.nextInt(upperBound: candidates.count)]
            switch try PublicPicPacSearchModel.place(state, at: cell) {
            case .win:
                return Rollout(reward: Double(sign), depth: depth)
            case .draw:
                return Rollout(reward: 0, depth: depth)
            case let .chance(chance):
                state = try PublicPicPacSearchModel.draw(chance, symbol: sample(chance))
                sign = -sign
                depth += 1
            }
        }
    }

    private func sample(_ chance: PicPacChanceState) throws -> Symbol {
        try random.nextInt(upperBound: chance.total) < chance.remainingX ? .x : .o
    }

    private func finish(
        path: [Stats],
        reward: Double,
        depth: Int,
        created: Int
    ) -> SimulationResult {
        for stats in path {
            stats.visits += 1
            stats.valueSum += reward
        }
        return SimulationResult(depth: depth, created: created)
    }

    private func bestFinalEdge(_ root: DecisionNode) -> ActionEdge? {
        var best: ActionEdge?
        for cell in root.edgeOrder {
            guard let edge = root.edges[cell] else { continue }
            guard let current = best else {
                best = edge
                continue
            }
            if edge.visits > current.visits
                || (edge.visits == current.visits && edge.mean > current.mean)
                || (edge.visits == current.visits
                    && edge.mean == current.mean
                    && moveRank(edge.cell) < moveRank(current.cell)) {
                best = edge
            }
        }
        return best
    }

    private func moveRank(_ cell: Cell) -> Int {
        switch cell.index {
        case 4: 0
        case 0: 1
        case 2: 2
        case 6: 3
        case 8: 4
        default: 5 + cell.index
        }
    }
}

private class Stats {
    var visits = 0
    var valueSum = 0.0
}

private final class DecisionNode: Stats {
    let state: PicPacDecisionState
    let rootSign: Int
    var untried: [Cell]
    var edges: [Cell: ActionEdge] = [:]
    var edgeOrder: [Cell] = []

    init(state: PicPacDecisionState, rootSign: Int) {
        self.state = state
        self.rootSign = rootSign
        untried = state.legalCells
    }
}

private final class ActionEdge: Stats {
    let cell: Cell
    var children: [Symbol: DecisionNode] = [:]
    var mean: Double { visits == 0 ? 0 : valueSum / Double(visits) }

    init(cell: Cell) {
        self.cell = cell
    }
}

private struct Rollout {
    let reward: Double
    let depth: Int
}

private struct SimulationResult {
    let depth: Int
    let created: Int
}
